cmd/cstor-sidecar/ccrd/controller-cstor.go

```diff
@@ -0,0 +1,340 @@
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		glog.Infof("Successfully synced '%s'", q.key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		runtime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the spcUpdated resource
+// with the current status of the resource.
+func (c *Controller) syncHandler(key, operation string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the spcUpdated resource with this namespace/name
+	if namespace == "" {
+		namespace = "default"
+	}
+
+	//spcUpdated, err := c.spcLister.StoragePoolClaims(namespace).Get(name)
+	_ = namespace
+	cstorCrdUpdated, err := c.clientset.OpenebsV1alpha1().CstorCrds(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		// The spc resource may no longer exist, in which case we stop
+		// processing.
+		if errors.IsNotFound(err) {
+			runtime.HandleError(fmt.Errorf("cstorCrdUpdated '%s' in work queue no longer exists", key))
+			return nil
+		}
+
+		return err
+	}
+
+	switch operation {
+	case "add":
+		fmt.Println("added event")
+		cachefile := "cachefile=" + cstorCrdUpdated.Spec.Zpool.Cachefile
+
+		cmdimport := exec.Command("zpool", "import", "-c",
+			cstorCrdUpdated.Spec.Zpool.Cachefile,
+			cstorCrdUpdated.Spec.Zpool.Poolname)
+		stdoutStderrImport, err := cmdimport.CombinedOutput()
+		if err != nil {
+			fmt.Println("err: ", err)
+			fmt.Println("stdoutStderr: ", string(stdoutStderrImport))
+		}else {
+			fmt.Println("Importing Successful")
+			return nil
+		}
+
+		cmd1 := exec.Command("zpool", "create", "-f", "-o", cachefile,
+			cstorCrdUpdated.Spec.Zpool.Poolname,
+			cstorCrdUpdated.Spec.Zpool.DiskPath)
+		fmt.Println("cmd1 : ", cmd1)
+		//res, err := cmd1.Output()
+		stdoutStderr, err := cmd1.CombinedOutput()
+		if err != nil {
+			fmt.Println("err: ", err)
+			fmt.Println("stdoutStderr: ", string(stdoutStderr))
+			return err
+		} else {
+			var compression string
+			var Readonly string
+			logbias := "logbias=" + cstorCrdUpdated.Spec.Zpool.Zfs.Logbias
+			copies := "copies=" + strconv.Itoa(cstorCrdUpdated.Spec.Zpool.Zfs.Copies)
+			sync := "sync=" + cstorCrdUpdated.Spec.Zpool.Zfs.Sync
+			fullvolname := cstorCrdUpdated.Spec.Zpool.Poolname + "/" + cstorCrdUpdated.Spec.Zpool.Zfs.Volname
+			fmt.Println("stdoutStderr: ", string(stdoutStderr))
+			if cstorCrdUpdated.Spec.Zpool.Zfs.Compression == true {
+				compression = "compression=on"
+			} else {
+				compression = "compression=off"
+			}
+			if cstorCrdUpdated.Spec.Zpool.Zfs.Readonly == true {
+				Readonly = "readonly=on"
+			} else {
+				Readonly = "readonly=off"
+			}
+			cmd2 := exec.Command("zfs", "create", "-s", "-b",
```

Comment:
I am not sure how backward & forward compatibility will be handled? How do we handle the cases, where zfs exposes new options that mandates change of spc attributes & hence version. 

Having a struct e.g. zpool, etc might handle these version cases. With newer versions, zpoolv1 or zpoolv2 structs can be created. This also implies the use of interfaces that can be used to use a particular struct.

Reply:
we dont need to change binary name from zpoolv1 or zpoolv2 for forward and backward compatibility. We can change the package name openebsv1alpha1 of autogenerated code which will reduce the effort and time and facilitates compatibility.